### ISSUE TITLE
Reporting of exceptions and error code

### ISSUE THREAD

**Issue by Stagno** [MeteoSwiss-APN/dawn#791]
When an exception occurs in dawn, a generic error is reported:
https://github.com/MeteoSwiss-APN/dawn/blob/8892214fb588d3da2f5542fec6a518769004576d/dawn/src/dawn/Compiler/DawnCompiler.cpp#L472-L475
but no information about the exception.
For development purposes we should probably let the exception through so that we can investigate it (and also let the debugger stop at the point where it's thrown).

Also, when this happens, gtclang prints "Compilation finished successfully" and returns a 0 exit code.

To reproduce, just throw an exception inside a pass.

**Comment by havogt** [comment]
At this stage of the project, I would just remove the try-catch block. Or if you want this "nice error", I would print the message and re-throw.

**Comment by Stagno** [comment]
But if you re-throw gdb will stop only at the point where you re-throw, right?

**Comment by havogt** [comment]
You are right...

**Comment by Stagno** [comment]
We need the `finally` like in Java! :)

**Comment by jdahm** [comment]
Yes, this has been on my mind in the optimizer refactor. My vote is to allow exceptions throughout the code (so passes could throw). Do you agree?

**Comment by Stagno** [comment]
Yes
